### PR TITLE
Allow attaching files to TestResults

### PR DESCRIPTION
**Pull request by ChrisMaddock**
This is the framework functionality required for #1670. (Adding this PR to the milestone separately, as I want to track runner implementation with the issue.)

Specific design points:

- API is `TestContext.AddTestAttachment(string filepath, string description = null)` Are we happy with that?

- I forced all files to be absolute paths - based on the logic that it's the runner that decides where the test result XML will end up, I believe? This is the same approach MSTest takes.

What do people think? I've also added some line comments on things to check.

### CODE REVIEW

File: src/NUnitFramework/framework/Internal/TestExecutionContext.cs [hunk 1]
```diff
@@ -493,6 +502,28 @@ public void AddFormatter(ValueFormatterFactory formatterFactory)
...
+        /// <param name="description">Optional description of attachment</param>
+        public void AddTestAttachment(string filepath, string description = null)
+        {
+            Guard.ArgumentNotNull(filepath, nameof(filepath));
+            Guard.ArgumentValid(filepath.IndexOfAny(Path.GetInvalidPathChars()) == -1, 
+                $"Test attachment filepath contains invalid path characters. {filepath}", nameof(filepath));
+
+            if (!Path.IsPathRooted(filepath))
+                filepath = Path.Combine(TestContext.CurrentContext.WorkDirectory, filepath);
+
```
Comment (ChrisMaddock) [inline src/NUnitFramework/framework/Internal/TestExecutionContext.cs]: Is this the correct way to build an absolute file path? I wanted to use `Path.GetFullPath()` - but it's not available in PCL.

File: src/NUnitFramework/tests/nunit.framework.tests-3.5.csproj [hunk 2]
```diff
@@ -122,6 +123,7 @@
     <Compile Include="Internal\Results\TestResultOutputTests.cs" />
     <Compile Include="Internal\Results\TestResultSuccessTests.cs" />
     <Compile Include="Internal\Results\TestResultTests.cs" />
+    <Compile Include="Internal\Results\TestResultWarningTests.cs" />
```
Comment (ChrisMaddock) [inline src/NUnitFramework/tests/nunit.framework.tests-3.5.csproj]: This is unrelated - just noticed this file was missing from all except the 2.0 build.

**Comment by ChrisMaddock** [comment]
Regarding file location - MSTest also copies the file - but has a fully fledged directory structure to support that.

This currently just references whatever the user passes in - meaning the link will be broken if the results are transferred between machines. Do we want to go any further, and (e.g.) copy the files to a directory next to the result xml? I wasn't entirely sure of the advantage, given a) potential copy permission errors and b) still requiring the user to copy both files and results, if that was the case.

I'm tempted to think the simple option will be ok, until such a point as someone has a compelling argument to change it?

**Comment by CharliePoole** [comment]
@ChrisMaddock What do you mean by "I want to track runner implementation with the issue?"

**Comment by ChrisMaddock** [comment]
@CharliePoole - There's a number of people following issue #1670. Rather than close it, and leave them to track down the related runner issues, which need to be completed before they can use the functionality, I've converted it to an Epic, and will add issues for the different runners, once we've decided what to do for each.

The side effect of that, is that there's no dedicated issue for just the framework component of #1670 - so I added the PR to the Epic.

**Comment by CharliePoole** [comment]
I'm glad you took this on as it's a needed feature. I do need a bit of general explanation before doing a detailed review.

You are adding an entry to TestExecutionContext for attachments and then copying them to the result. Is there any reason you don't want to simply add them to the result directly? This would be analogous to how we add text output to the result.

In the TestExecutionContext constructor, you are copying the attachments from each higher level to the lower level. Is there a specific reason for that? Won't that mean the attachments are duplicated all the way down the tree of results?

**Comment by ChrisMaddock** [comment]
> You are adding an entry to TestExecutionContext for attachments and then copying them to the result. Is there any reason you don't want to simply add them to the result directly? This would be analogous to how we add text output to the result.

I hadn't realised the TestResult was created at the start of the test - I followed where StartTime was set, which was in WorkItem.WorkItemComplete(). This sounds like a much better idea.

> In the TestExecutionContext constructor, you are copying the attachments from each higher level to the lower level. Is there a specific reason for that? Won't that mean the attachments are duplicated all the way down the tree of results?

Nope - that's just me not looking into things properly. Sorry.

It's getting on a bit UK-time, will make the changes tomorrow. Thanks for those!

**Comment by CharliePoole** [comment]
@ChrisMaddock Great! I'll wait till you update to do a detailed review.

File: src/NUnitFramework/tests/Internal/TestExecutionContextTests.cs [hunk 3]
```diff
@@ -907,6 +918,65 @@ public void SetAndRestoreValueFormatter()
...
+            Assert.That(() => context.AddTestAttachment("NotAFile.txt"), Throws.InstanceOf<FileNotFoundException>());
+        }
+
+        public TestExecutionContext CreateFakeContext()
+        {
+            var fakeResult = TestUtilities.Fakes.GetTestMethod(this, "FakeMethod").MakeTestResult();
+            var context = new TestExecutionContext { CurrentResult = fakeResult };
+            return context;
+        }
+
```
Comment (ChrisMaddock) [inline src/NUnitFramework/tests/Internal/TestExecutionContextTests.cs]: `CreateFakeContext()` is pretty ugly. Am I missing a better way to create temporary context to test with? I didn't want to attach anything to the actual TestExecutionContext.

**Comment by ChrisMaddock** [comment]
Initial changes made, and ready for review. Thanks Charlie.

File: src/NUnitFramework/tests/Internal/TestExecutionContextTests.cs (hunk 3 above)
Comment (CharliePoole) [inline src/NUnitFramework/tests/Internal/TestExecutionContextTests.cs]: It's easy to create a context, harder to populate it with meaningful values. In this case, we needed a context with a valid result. The alternative is to use NSubstitute but I think you'll be doing a lot of setup if you do that.

File: src/NUnitFramework/framework/Internal/Results/TestResult.cs [hunk 4]
```diff
@@ -268,7 +283,7 @@ internal set
                 InternalAssertCount = value;
             }
         }
-
+
```
Comment (jnm2) [inline src/NUnitFramework/framework/Internal/Results/TestResult.cs]: I really hate to nitpick whitespace. I like these tools:
VS: https://marketplace.visualstudio.com/items?itemName=MadsKristensen.TrailingWhitespaceVisualizer
VS Code: https://marketplace.visualstudio.com/items?itemName=shardulm94.trailing-spaces with `"trailing-spaces.trimOnSave": true`

File: src/NUnitFramework/framework/Internal/Results/TestResult.cs [hunk 5]
```diff
@@ -701,6 +719,26 @@ private ResultState AssertionStatusToResultState(AssertionStatus status)
         }
 
         /// <summary>
+        /// Adds a attachments element to a node and returns it.
+        /// </summary>
+        /// <param name="targetNode">The target node.</param>
+        /// <returns>The new attachments element.</returns>
+        private TNode AddAttatchmentsElement(TNode targetNode)
```
Comment (jnm2) [inline src/NUnitFramework/framework/Internal/Results/TestResult.cs]: Nit: should be `AddAttachmentsElement`

File: src/NUnitFramework/framework/Internal/TestAttachment.cs [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+    /// <summary>
+    /// The TestAttachment class represents a file attached to a TestResult,
+    /// with an optional description.
+    /// </summary>
+    internal class TestAttachment
+    {
+        /// <summary>
+        /// Absolute filepath to attachment file
+        /// </summary>
+        public string Filepath { get; }
```
Comment (jnm2) [inline src/NUnitFramework/framework/Internal/TestAttachment.cs]: I've never seen this as one word. Can it be `FilePath`?

File: src/NUnitFramework/framework/Internal/TestAttachment.cs [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+        /// Optional user specifed description of attachment
+        /// </summary>
+        public string Description { get; }
+
+        /// <summary>
+        /// Creates a TestAttachment class to represent a file attached to a test result.
+        /// </summary>
+        /// <param name="filepath">Absolute filepath to attachment file</param>
+        /// <param name="description">Optional user specifed description of attachment</param>
+        public TestAttachment(string filepath, string description)
```
Comment (jnm2) [inline src/NUnitFramework/framework/Internal/TestAttachment.cs]: Same; I've never seen this as one word. Can it be `filePath`?

File: src/NUnitFramework/framework/Internal/TestExecutionContext.cs [hunk 8]
```diff
@@ -493,6 +494,28 @@ public void AddFormatter(ValueFormatterFactory formatterFactory)
             CurrentValueFormatter = formatterFactory(CurrentValueFormatter);
         }
 
+        /// <summary>
+        /// Attach a file to the current test result
+        /// </summary>
+        /// <param name="filepath">Filepath to attachment. Can be absolute, or relative to working directory.</param>
+        /// <param name="description">Optional description of attachment</param>
+        public void AddTestAttachment(string filepath, string description = null)
```
Comment (jnm2) [inline src/NUnitFramework/framework/Internal/TestExecutionContext.cs]: Same; I've never seen this as one word. Can it be filePath?

File: src/NUnitFramework/tests/Internal/Results/TestResultAttachmentTests.cs [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+namespace NUnit.Framework.Internal.Results
+{
+    public class TestResultAttachmentTests
+    {
+        private TestResult _result;
+
+        [SetUp]
+        public void SetUp()
+        {
+            _result = TestUtilities.Fakes.GetTestMethod(this, "FakeMethod").MakeTestResult();
```
Comment (jnm2) [inline src/NUnitFramework/tests/Internal/Results/TestResultAttachmentTests.cs]: `nameof(FakeMethod)` 😀

File: src/NUnitFramework/tests/Internal/TestExecutionContextTests.cs [hunk 10]
```diff
@@ -43,28 +44,38 @@ namespace NUnit.Framework.Internal
...
+        private string _tempFilePath;
+        private DateTime _fixtureCreateTime = DateTime.UtcNow;
+        private long _fixtureCreateTicks = Stopwatch.GetTimestamp();
         
         [OneTimeSetUp]
         public void OneTimeSetUp()
         {
             _fixtureContext = TestExecutionContext.CurrentContext;
             _fixtureResult = _fixtureContext.CurrentResult.ResultState;
+            _tempFilePath = Path.Combine(TestContext.CurrentContext.WorkDirectory, "NUnitTests.tmp");
```
Comment (jnm2) [inline src/NUnitFramework/tests/Internal/TestExecutionContextTests.cs]: Replace `"NUnitTests.tmp"` with `TempFileName`?

File: src/NUnitFramework/tests/Internal/TestExecutionContextTests.cs [hunk 11]
```diff
@@ -907,6 +918,65 @@ public void SetAndRestoreValueFormatter()
...
+        public void NoneExistandFileThrowsFileNotFoundException()
+        {
+            var context = CreateFakeContext();
+            Assert.That(() => context.AddTestAttachment("NotAFile.txt"), Throws.InstanceOf<FileNotFoundException>());
+        }
+#endif
+
+        public TestExecutionContext CreateFakeContext()
+        {
+            var fakeResult = TestUtilities.Fakes.GetTestMethod(this, "FakeMethod").MakeTestResult();
```
Comment (jnm2) [inline src/NUnitFramework/tests/Internal/TestExecutionContextTests.cs]: `nameof(FakeMethod)` 😀

File: src/NUnitFramework/framework/Internal/Results/TestResult.cs (hunk 5 above)
Comment (ChrisMaddock) [inline src/NUnitFramework/framework/Internal/Results/TestResult.cs]: I never could spell that word! Thanks!

File: src/NUnitFramework/tests/Internal/TestExecutionContextTests.cs (hunk 10 above)
Comment (ChrisMaddock) [inline src/NUnitFramework/tests/Internal/TestExecutionContextTests.cs]: Thanks. I had to do a bit of shuffling here to get things playing nice with PCL - must have missed that. 🙂

File: src/NUnitFramework/framework/Internal/TestAttachment.cs (hunk 6 above)
Comment (ChrisMaddock) [inline src/NUnitFramework/framework/Internal/TestAttachment.cs]: I Googled, and you're right. I'll change throughout - thanks!

**Comment by ChrisMaddock** [comment]
Thanks @jnm2. I'll wait for comments from others, then fix everything in one go. 😄

File: src/NUnitFramework/framework/Internal/Results/TestResult.cs (hunk 4 above)
Comment (rprouse) [inline src/NUnitFramework/framework/Internal/Results/TestResult.cs]: @jnm2, LOL, nice to see other developers as OCD as me 😄

File: src/NUnitFramework/framework/Internal/Results/TestResult.cs [hunk 12]
```diff
@@ -701,6 +719,26 @@ private ResultState AssertionStatusToResultState(AssertionStatus status)
...
+        {
+            TNode attachmentsNode = targetNode.AddElement("attachments");
+
+            foreach (var attachment in TestAttachments)
+            {
+                var attachmentNode = attachmentsNode.AddElement("attachment", attachment.Filepath);
+
+                if (attachment.Description != null)
+                    attachmentNode.AddAttribute("description", attachment.Description);
```
Comment (rprouse) [inline src/NUnitFramework/framework/Internal/Results/TestResult.cs]: Description is a user supplied string that may be long or require CDATA encoding. Based on that, we should probably make it a child element rather than an attribute and CDATA encode it.

```C#
var attachmentNode = attachmentsNode.AddElement("attachment");
attachmentNode.AddElement("filename", attachment.Filepath);

if (attachment.Description != null)
    attachmentNode.AddElementWithCDATA("description", attachment.Description);
```

Tests will need to be changed accordingly.

File: src/NUnitFramework/framework/Internal/TestExecutionContext.cs (hunk 1 above)
Comment (rprouse) [inline src/NUnitFramework/framework/Internal/TestExecutionContext.cs]: Looks good to me, although I think `Path.GetFullPath()` will be available once I get the .NET Standard version working.

File: src/NUnitFramework/framework/Internal/Results/TestResult.cs (hunk 12 above)
Comment (ChrisMaddock) [inline src/NUnitFramework/framework/Internal/Results/TestResult.cs]: Sure. 

I did consider this - the current AddAttribute goes through an ExcapeXmlCharacters() - which should sufficiently encode any user string. If length is a concern for readability however, I can change it over. Will do that with the rest. 🙂

**Comment by ChrisMaddock** [comment]
Thanks both. @CharliePoole - you said earlier you were planning to come back to this - so I'll hold on for your feedback, then deal with everything in one go.

File: src/NUnitFramework/framework/Internal/Results/TestResult.cs (hunk 4 above)
Comment (CharliePoole) [inline src/NUnitFramework/framework/Internal/Results/TestResult.cs]: OTOH, if VS doesn't really care...

File: src/NUnitFramework/framework/Internal/Results/TestResult.cs [hunk 13]
```diff
@@ -80,7 +80,13 @@ public abstract class TestResult : ITestResult
...
         private string _stackTrace;
 
-        private List<AssertionResult> _assertionResults = new List<AssertionResult>();
+        private readonly List<AssertionResult> _assertionResults = new List<AssertionResult>();
+
+        /// <summary>
+        /// TestAttachments attached to result.
+        /// Only exposed for testing.
+        /// </summary>
+        internal ICollection<TestAttachment> TestAttachments { get; } = new List<TestAttachment>();
```
Comment (CharliePoole) [inline src/NUnitFramework/framework/Internal/Results/TestResult.cs]: Working my way down... I can't see why this would not be `private readonly List`. Maybe it will become clear as I go. 😄

File: src/NUnitFramework/framework/Internal/Results/TestResult.cs (hunk 5 above)
Comment (CharliePoole) [inline src/NUnitFramework/framework/Internal/Results/TestResult.cs]: Spelling is not a nitpick. Just whitespace. 😄

File: src/NUnitFramework/framework/Internal/Results/TestResult.cs (hunk 12 above)
Comment (CharliePoole) [inline src/NUnitFramework/framework/Internal/Results/TestResult.cs]: Amazing how things grow! I didn't imagine we would have a description at all at the start but if we have one I think @rprouse is right.

File: src/NUnitFramework/framework/Internal/TestExecutionContext.cs (hunk 1 above)
Comment (CharliePoole) [inline src/NUnitFramework/framework/Internal/TestExecutionContext.cs]: Actually `Path.GetFullPath()` would be incorrect. We promise to store all output files relative to our WorkDirectory, which is __not__ the current directory. The choice of name for the WorkDirectory, may have been a bad one, since it sounds a bit like the working directory, i.e. cwd, but I believe we are stuck with it.

A more basic question... why do we want to make this absolute at all? Why not just use what the user gives us?

File: src/NUnitFramework/framework/Internal/TestExecutionContext.cs [hunk 14]
```diff
@@ -493,6 +494,28 @@ public void AddFormatter(ValueFormatterFactory formatterFactory)
...
+            Guard.ArgumentNotNull(filepath, nameof(filepath));
+            Guard.ArgumentValid(filepath.IndexOfAny(Path.GetInvalidPathChars()) == -1, 
+                $"Test attachment filepath contains invalid path characters. {filepath}", nameof(filepath));
+
+            if (!Path.IsPathRooted(filepath))
+                filepath = Path.Combine(TestContext.CurrentContext.WorkDirectory, filepath);
+
+#if !PORTABLE
+            if (!File.Exists(filepath))
+                throw new FileNotFoundException("Test attachment filepath could not be found.", filepath);
```
Comment (CharliePoole) [inline src/NUnitFramework/framework/Internal/TestExecutionContext.cs]: Why do we care that the file doesn't exist? Maybe the user asked to add the attachment before saving it. Maybe some third party software is expected to add it later. If we are not doing anything with the file, why not just record what the user tells us to record?

File: src/NUnitFramework/framework/TestContext.cs [hunk 15]
```diff
@@ -295,6 +295,16 @@ public static void AddFormatter(ValueFormatterFactory formatterFactory)
         }
 
         /// <summary>
+        /// Attach a file to the current test result
+        /// </summary>
+        /// <param name="filepath">Relative or absolute filepath to attachment</param>
+        /// <param name="description">Optional description of attachment</param>
+        public static void AddTestAttachment(string filepath, string description = null)
+        {
+            TestExecutionContext.CurrentContext.AddTestAttachment(filepath, description);
```
Comment (CharliePoole) [inline src/NUnitFramework/framework/TestContext.cs]: Why the extra level of indirection? In other code, TestContext just accesses `TestExecutionContext.CurrentContext.CurrentResult`  directly.

File: src/NUnitFramework/framework/Internal/Results/TestResult.cs (hunk 13 above)
Comment (jnm2) [inline src/NUnitFramework/framework/Internal/Results/TestResult.cs]: Comment says "Only exposed for testing." but `readonly` sounds good.

Comment (CharliePoole) [inline src/NUnitFramework/framework/Internal/Results/TestResult.cs]: Ah! I didn't notice the comment. So this field is not the private backing field for a public property then? That seems a bit odd, since it leaves a part of the TestResult inaccessible to clients.

Comment (ChrisMaddock) [inline src/NUnitFramework/framework/Internal/Results/TestResult.cs]: You're right. I had been keeping this as an 'add-only' collection, but it would be more in-keeping with the rest of the TestResult class if this was exposed to the user in a ReadOnlyCollection. I'll expose this list and the TestAttachment type.

File: src/NUnitFramework/framework/Internal/TestExecutionContext.cs (hunk 14 above)
Comment (ChrisMaddock) [inline src/NUnitFramework/framework/Internal/TestExecutionContext.cs]: This was a point I went back and forth on. 🙂 My thinking was that adding a non-existant file is far more likely to occur through user-error, rather than intention, and you'd want to know about that ASAP. I imagine a common use case may be where people only refer to attachments when the test has failed - so broken paths may only be discovered 6 months after writing, when the test first fails, and it's too late.

The opposite arguments e.g. 'adding an attachment before saving it', mostly seem like things that could be managed by the user. (e.g. Don't add an attachment before you've created it!) That seems like fairly bad practice anyway, what if the creation of the attachment later went wrong, and the file couldn't be written? We could always add a 'don't check file existence' overload in the future if such cases arose.

Not that it matters, but MSTest checks if the File Exists. @rprouse @jnm2 - what do you guys think?

File: src/NUnitFramework/framework/TestContext.cs (hunk 15 above)
Comment (ChrisMaddock) [inline src/NUnitFramework/framework/TestContext.cs]: I can change that if this ways preferred. I get a little mixed up about logic should live between TestContext and TestExecutionContext - I'd followed how `AddFormatter()` was written.

This is also more possible now I'm no longer storing an intermediate list of attachments on TestExecutionContext. 😄

File: src/NUnitFramework/framework/Internal/TestExecutionContext.cs (hunk 1 above)
Comment (ChrisMaddock) [inline src/NUnitFramework/framework/Internal/TestExecutionContext.cs]: Thanks for the explanation! 

This was something else I went back and forth on. I wanted to make it absolute, as I had previously thought:
- The attachments should always be relative the the results xml - as the file paths written there should be correct to the files current location.
- The runner controls where the results xml is written, meaning that the runner could theoretically move where the results ended up.

What I didn't realise is that WorkDirectory wasn't the working directory. 😄 So will any results xml file always end up in TestContext.WorkDirectory? Or is that a runner setting, could TestResults be written anywhere, like I'd first thought?

I did think full paths would be easier to resolve should the user move the result xml file around for any CI etc. Is there a disadvantage to writing full paths, if we already define that they're relative to WorkDirectory? I guess if they remain relative, and you were adding all your attachments to the same subdir, you could move the directory as a whole?

**Comment by ChrisMaddock** [comment]
Thanks @CharliePoole! I've replied to the inline comments, the remaining ones...

> I would have tested this using a new class in the testdata-assembly. That would eliminate all worries about test context, execution context, etc. We would simply be looking at the result and it's fields.

I'll try it. 🙂 

> There is no test of TestContext
No - I'll move the tests currently in TestExecutionContext to use TestContext, once I move everything out of there.

> There are no new tests of TestResult, particularly to verify that it creates XML correctly.
Is this not covered by TestResultAttachmentTests?

Will hopefully make the changes this weekend - thanks all!

File: src/NUnitFramework/framework/Internal/TestExecutionContext.cs (hunk 14 above)
Comment (CharliePoole) [inline src/NUnitFramework/framework/Internal/TestExecutionContext.cs]: I guess checking for existence makes sense in general.

File: src/NUnitFramework/framework/Internal/TestExecutionContext.cs (hunk 1 above)
Comment (CharliePoole) [inline src/NUnitFramework/framework/Internal/TestExecutionContext.cs]: The runner passes Work directory to the framework to be used in exactly this circumstance. The XML result file is saved there. 
In the case of the console runner, it can be set by a command-line argument.

Comment (ChrisMaddock) [inline src/NUnitFramework/framework/Internal/TestExecutionContext.cs]: Ok, cool. I'll remove the stuff that meddles with paths in that case - if we have the concrete definition that paths will always be relative to WorkDirectory/results location. Thanks.

File: src/NUnitFramework/framework/Internal/TestExecutionContext.cs (hunk 14 above)
Comment (jnm2) [inline src/NUnitFramework/framework/Internal/TestExecutionContext.cs]: Typically, if you're going to do anything else with the file other than verify that it exists, the existence check is meaningless because it's race condition prone. To have meaningful handling you'd have to catch FileNotFoundException because it's the only atomic API.

> I imagine a common use case may be where people only refer to attachments when the test has failed - so broken paths may only be discovered 6 months after writing, when the test first fails, and it's too late.

But this is a great point and would be very helpful.

**Comment by ChrisMaddock** [comment]
Review changes made - this is ready for a second look. 🙂 

Only thing to note, is that this contains `if #PORTABLE` symbols, so either this or #2174 will need adjusting, depending which is merged first.

**Comment by oznetmaster** [comment]
I commented on the commit.  Probably the wrong place, again :(
